core/tests/compile_error.rs

```diff
@@ -1,41 +1,101 @@
 //! Integration tests for GDLK that expect compile errors. The programs in
 //! these tests should all fail during compilation.
 
-use gdlk::{Compiler, HardwareSpec, Valid};
+use gdlk::{ast::LangValue, Compiler, HardwareSpec, Valid};
 
 /// Compiles the program for the given hardware, expecting compile error(s).
 /// Panics if the program compiles successfully, or if the wrong set of
 /// errors is returned.
-fn expect_compile_errors(
-    hardware_spec: HardwareSpec,
-    src: &str,
-    expected_errors: &[&str],
-) {
-    // Compile from hardware+src
-    let actual_errors =
-        Compiler::compile(src.into(), Valid::validate(hardware_spec).unwrap())
-            .unwrap_err();
-    assert_eq!(format!("{}", actual_errors), expected_errors.join("\n"));
+macro_rules! assert_compile_errors {
+    ($hw_spec:expr, $src:expr, $expected_errors:expr $(,)?) => {
+        // Compile from hardware+src
+        let actual_errors: Vec<String> =
+            Compiler::compile($src.into(), Valid::validate($hw_spec).unwrap())
+                .unwrap_err()
+                .errors()
+                .iter()
+                .map(|err| format!("{}", err))
+                .collect();
+        let strs: Vec<&str> =
+            actual_errors.iter().map(String::as_str).collect();
+        assert_eq!(strs.as_slice(), $expected_errors);
+    };
+}
```

Comment:
I switched to macros cause then the line that triggers the error is in the test, instead of always just pointing to the same line in the helper function.